### PR TITLE
kubeadm: support image repository format validation

### LINKED ISSUE #2758: Expected imageRepository field to be validated
## What keywords did you search in kubeadm issues before filing this one?

validate 
imageRepository

## Is this a BUG REPORT or FEATURE REQUEST?
Bug report

## Versions

**kubeadm version** (use `kubeadm version`): v1.25.0

**Environment**:
- **Kubernetes version** (use `kubectl version`): N/A
- **Cloud provider or hardware configuration**: N/A
- **OS** (e.g. from /etc/os-release): N/A
- **Kernel** (e.g. `uname -a`): N/A
- **Container runtime (CRI)** (e.g. containerd, cri-o): N/A
- **Container networking plugin (CNI)** (e.g. Calico, Cilium): N/A
- **Others**: N/A

## What happened?

I modified a kind kubeadm configuration as follows:

```yaml
apiServer:
  certSANs:
  - localhost
  - 127.0.0.1
  extraArgs:
    runtime-config: ""
apiVersion: kubeadm.k8s.io/v1beta3
clusterName: kind
controlPlaneEndpoint: kind-control-plane:6443
dns:
  imageRepository: |
    registry.contoso.com/kubernetes
      s
```

the key bit being

```yaml
dns:
  imageRepository: |
    registry.contoso.com/kubernetes
      s
```

This was then passed into `kubeadm init --config blah.yaml`, and kubeadm proceeded until preflight with

```
[preflight] You can also perform this action in beforehand using 'kubeadm config images pull'
error execution phase preflight: [preflight] Some fatal errors occurred:
        [ERROR ImagePull]: failed to pull image registry.contoso.com/kubernetes
  s
/coredns:v1.9.3: output: E0921 13:07:35.156113    6667 remote_image.go:238] "PullImage from image servic
e failed" err="rpc error: code = Unknown desc = failed to parse image reference \"registry.contoso.com/kubernetes\\n  s\\n/coredns:v1.9.3\": invalid reference format" image=<
        registry.contoso.com/kubernetes
          s
        /coredns:v1.9.3
 >
```

## What you expected to happen?

kubeadm to throw a validation error.

See also https://github.com/kubernetes-sigs/cluster-api/issues/7257 

## How to reproduce it (as minimally and precisely as possible)?

See above.

### PR DESCRIPTION
**Pull request by SataQiu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
kubeadm: support image repository format validation

Sample output：
```bash
./kubeadm init --config /tmp/kubeadm.config.yaml --dry-run
dns.imageRepository: Invalid value: "registry.contoso.com/kubernetes\n  s\n": invalid image repository format
```

#### Which issue(s) this PR fixes:

Fixes kubernetes/kubeadm#2758

#### Does this PR introduce a user-facing change?

```release-note
kubeadm: support image repository format validation
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 1]
```diff
@@ -638,3 +649,15 @@ func ValidateSocketPath(socket string, fldPath *field.Path) field.ErrorList {
 
 	return allErrs
 }
+
+// ValidateImageRepository validates format of image repository
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:653]:
```suggestion
// ValidateImageRepository validates the image repository format
```

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 2]
```diff
@@ -638,3 +649,15 @@ func ValidateSocketPath(socket string, fldPath *field.Path) field.ErrorList {
...
+func ValidateImageRepository(imageRepository string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	dummyImage := fmt.Sprintf("%s/%s:%s", imageRepository, "dummy", "1.0")
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:657]: i think we should change the word "dummy" to just "image".
i remember some complains about it.

**Review by neolit123 — COMMENTED** [review]
/triage accepted
thanks @SataQiu

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 3]
```diff
@@ -66,6 +67,7 @@ func ValidateClusterConfiguration(c *kubeadm.ClusterConfiguration) field.ErrorLi
 	allErrs = append(allErrs, ValidateAbsolutePath(c.CertificatesDir, field.NewPath("certificatesDir"))...)
 	allErrs = append(allErrs, ValidateFeatureGates(c.FeatureGates, field.NewPath("featureGates"))...)
 	allErrs = append(allErrs, ValidateHostPort(c.ControlPlaneEndpoint, field.NewPath("controlPlaneEndpoint"))...)
+	allErrs = append(allErrs, ValidateImageRepository(c.ImageRepository, field.NewPath("imageRepository"))...)
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:70]: if the field is empty would the validation function pass?

**Review by neolit123 — COMMENTED** [review]
/lgtm
/hold

please cancel the hold if this is good to go

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 3 above)
Comment (SataQiu) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go:70]: Yes, we always do validation after defaulting.

**Comment by SataQiu** [comment]
Let's wait for more review 
cc @randomvariable @pacoxu

**Comment by pacoxu** [comment]
/lgtm

BTW, should we backport it to old releases so that CAPI can benefit from this with earlier releases?

**Comment by neolit123** [comment]
> BTW, should we backport it to old releases so that CAPI can benefit from this with earlier releases?

well, technically it does not qualify for a backport because it is an improvement / feature and not a blocking bug fix.

**Comment by SataQiu** [comment]
it seems good to go
/hold cancel

**Pull request merged**